### ISSUE TITLE
Consider (de)modulating stft

### ISSUE THREAD

**Issue by OverLordGoldDragon** [librosa/librosa#1279]
Most STFT implementations, including this one, [do not](https://dsp.stackexchange.com/a/72590/50076) compute STFT per [standard formulation](https://en.wikipedia.org/wiki/STFT#/Discrete-time_STFT). Some use a "(de)modulated" variant, as in [this paper](https://arxiv.org/abs/1006.2533):

![image](https://user-images.githubusercontent.com/16495490/106966036-fed62a80-675d-11eb-9e9d-a62b7618b024.png)

This centers the windows at zero-phase of DFT bases. `librosa`'s implementation already has a shift (as do others), but it's much less favorable in one regard: numeric stability. For localized component / band analysis, results are extremely sensitive to inclusion or exclusion of rows, which has adverse effects on e.g. reconstruction and synchrosqueezing. Details [here](https://dsp.stackexchange.com/a/72590/50076).

The tweak is [quite simple](https://github.com/OverLordGoldDragon/ssqueezepy/blob/0.5.5/ssqueezepy/_stft.py#L106) - consider adding as option.

**Comment by bmcfee** [comment]
Sorry, but I don't see how this constitutes a "numeric stability" issue.  Is it not just a difference in phase, and if so, wouldn't this lead magnitudes (and relative phase differences from frame to frame) unchanged?  

Also, could you please provide a minimal working example of this being a problem?

(As an aside, if most implementations do not do something a certain way, it's questionable to refer to it as "standard formulation".)

**Comment by dpwe** [comment]
Centering can make a big difference when unwrapping phase along frequency,
but I don’t know what reassignment actually entails.

  DAn.

**Comment by OverLordGoldDragon** [comment]
By "standard formulation" I mean standard mathematical formula, which is indeed this:

![image](https://user-images.githubusercontent.com/16495490/107057743-35a75180-67ed-11eb-8b72-186ad028f41b.png)

Pasting relevant excerpt from [here](https://dsp.stackexchange.com/a/72590/50076); `modulated=False` is `librosa`'s behavior.

<hr>

<img src="https://i.stack.imgur.com/Jcg65.png">

"Modulated" (modified) sums well both directly and via `abs()`, while default creates a mess. Key is the last plots, showing `Sx = stft(x)` values for a single column. This is the direction we synchrosqueeze, or sum, along; summing directly, we alternate between locally close peaks of opposite signs, keeping the sum total small. Adding or removing just one peak can amount to doubling the net-sum _and_ flipping its sign.

To contrary, modulated is a lot smoother; no big deal if a few samples get misassigned. How's it work? For once, the answer's in `fft(fft(x))`; aside unimportant differences, DFT of DFT of `x` is `x`, and `x` is the windowed signal:

<img src="https://i.stack.imgur.com/9MOwO.png">

A _centered_ window thus corresponds to `fft(x)` that has high-frequency oscillations, whereas a window 'centered' about `n = 0` (wrapping around) bears low frequencies, thus the smooth plot.

**Comment by bmcfee** [comment]
> By "standard formulation" I mean standard mathematical formula, which is indeed this:

Yes, that is a common definition, but I would not consider it a "standard" since A) most implementations do not use it in that form, and B) there is no otherwise governing body defining it as such.  This is a minor point, but it could very easily lead to confusion among newbies finding this thread in google search results, so I want it clarified here.

On to your example: it still looks to me like the only difference is a circular shift of the windowed signal (after windowing), which should only amount to a difference in the phase spectrum, right?  I'm sure there are applications where this is helpful (per your synchrosqueezing example), but I'm not convinced that it's worth complicating the STFT API here if it can equivalently be handled as a post-processing of the STFT.  I'd be curious for @lostanlen 's input though.

**Comment by OverLordGoldDragon** [comment]
Key plots:

![image](https://user-images.githubusercontent.com/16495490/107062479-c7fe2400-67f2-11eb-96a5-419ab150c7bf.png)

This concerns any application working with _parts_ of `Sx` rather than whole, including reconstruction, ridge extraction, and reassignment. Adjacent rows have similar values but opposite signs, and nearly cancel in sum, so inclusion or exclusion of a single row drastically sways the end result.

**Comment by bmcfee** [comment]
> Adjacent rows have similar values but opposite signs, and nearly cancel in sum, so inclusion or exclusion of a single row can drastically sway the end result.

Yes, but this is because you're only looking at the real part and not treating the complex spectrum as complex numbers (for which "sign" isn't quite a coherent concept).  The magnitude spectrum should be the same with or without the fftshift, right?

**Comment by OverLordGoldDragon** [comment]
Magnitude exactly the same indeed, and any methods that operate on magnitude won't see a difference. Reconstruction (of a real input), however, deals with the real part, and reassignment with real and imaginary, rather than modulus.

**Comment by bmcfee** [comment]
Ok, I've turned this over a bit in my head, and have a couple more thoughts on this:

1. Sticking an `fftshift` inside the stft function will work, but it's somewhat inefficient as it invokes another array copy.  Instead, I'd suggest compensating for the half-frame delay afterward by doing an in-place broadcast multiply, something like : `D *= np.exp(-2j * np.pi * np.arange(n_fft // 2) * (n_fft // 2) / n_fft)`  (this might need a sign flip to be correct, haven't tested it yet).
2. Since the (de)modulation can be accomplished more efficiently by a post-processing step, I'm less inclined to include it into the API.  However...
3. One aspect that we haven't touched upon here is the interaction between frame length and window length.  In the current implementation, we do frequency over-sampling by padding the windowed frame (or, equivalently, padding the window for the target frame length).  This works well enough, but it means that the delay (and hence, phase) of the windowed signal will be different for different frame lengths, even with the same window length.  If we add this delay correction step, then we should get consistent phase behavior for varying frame lengths (assuming the window and hop lengths are consistent), and this seems like a win to me.  Not enough of a win to make it the default behavior, but I can see it as a nice addition.

**Comment by OverLordGoldDragon** [comment]
It's just the window shifted, which is of trivial computational expense, but sure one can optimize. For post-processing one's to rewrite all elements of `Sx` via multiply, which is lot more compute than pre-shifting the window. As for frame effects on phase, I don't see window shift introducing anything new, the relation remains `Sx_mod / Sx == (-1)**np.arange(...)` - or, exactly,

```python
Sx     = stft(x, modulated=False)
Sx_mod = stft(x, modulated=True)
Sx_mod/Sx == Sx/Sx_mod == np.hstack([(-1)**np.arange(len(Sx))[None].T]*Sx.shape[1])
```
Also I'm unsure on "modulated" vs "demodulated"; on one hand we're introducing explicit modulation, on another the modulation's always there and we're undoing it relative to `fft`. API-wise a shorter name's preferred, so I went with `modulated`.

**Comment by OverLordGoldDragon** [comment]
Nevermind on `Sx_mod/Sx`, I only considered the case `n_fft == len(x)`. I'll take a look.

**Comment by OverLordGoldDragon** [comment]
1. Implementing a `util.frame(.., modulated=True)` so values are seated correctly the first time, then `fftshift(fft_window)`, seems most efficient. As for in-place multiply, firstly `arange(n_fft//2) -> arange(n_fft//2+1)`, and the sign's right, but the `*(n_fft//2)` is `*(n_fft//2+1)` sometimes, and sometimes the whole thing's just a `(-1)**arange(...)`, depending on even/odd combination of `len(x), n_fft`. Haven't deduced exact relationship.
 2. Including it informs users the option exists, making it default suggests it's the better general option. Advantage over libraries that don't.
 3. Do you refer to `angle(fft(fft_window))`, and it in relation to `fftshift`ed? Both can be made zero-phase there. 

Also I'm unsure on "modulated" vs "demodulated"; on one hand we're introducing explicit modulation, on another the modulation's always there and we're undoing it relative to `fft`. API-wise a shorter name's preferred, so I went with `modulated`.

**Comment by OverLordGoldDragon** [comment]
Nicer example with pure tone and a time-localized window; real & imag components follow a bell and are consistent in sign along columns:

<details>
   <summary>code</summary>

```python
import numpy as np
import matplotlib.pyplot as plt
import librosa
import scipy.signal

x = np.cos(2*np.pi * 8 * np.linspace(0, 1, 128))
window = scipy.signal.windows.dpss(len(x), 12)

Sx = librosa.stft(x, window=window, hop_length=1, n_fft=128, modulated=True)

tkw = dict(weight='bold', fontsize=16, loc='left')
plt.imshow(Sx.real, cmap='bwr', aspect='auto')
plt.title("real(STFT)", **tkw)
plt.show()

plt.plot(Sx[:, 1].real)
plt.plot(Sx[:, 1].imag)
plt.title("STFT[:, 1], real, imag", **tkw)
plt.show()
```

</details>

![image](https://user-images.githubusercontent.com/16495490/107163189-a56e3580-69c1-11eb-832a-5866f583008c.png)

**Comment by bmcfee** [comment]
>     1. Implementing a `util.frame(.., modulated=True)` so values are seated correctly the first time, then `fftshift(fft_window)`, seems most efficient.

I don't think this is true.  In typical cases, `frame` does not create a new copy of the array, but fftshifting always will, so we'll be paying twice in memory consumption (and array copying).

In-place multiplying the output stft would still pay for the multiply operation (obviously), but it would not require a new array copy, so it should be much faster.


>     2\. Including it informs users the option exists, making it default suggests it's the better general option.

I think at this point, I'm persuaded that it's a worthwhile feature, but I don't think it should be the default.

**Comment by OverLordGoldDragon** [comment]
We'd only `fftshift(fft_window)`, which is a 1D array, not `y_frames`, which is `n_fft` times bigger in size. The window can also be spared of a copy, sure. Idea is, either way we use `util.frame` for same amount of array writes, so better to write it right the first time and spare any post-processing (like multiplying).

Fair re:default.

**Comment by bmcfee** [comment]
> We'd only fftshift `fft_window`, which is a 1D array, not `y_frames`, which is `n_fft` times bigger in size.

Your code example does both:


> Just added below to work with `librosa`:
> 
> ```python
> if modulated:
>     y_frames = np.fft.ifftshift(y_frames, axes=0)
>     fft_window = np.fft.ifftshift(fft_window)
> ```

If we shift the window but not the input frame, we don't get any of the benefits of windowing for reducing spectral leakage.  Discontinuities at the frame boundaries will persist because the shifted window will be large / far from zero at samples `0` and `n_fft-1`.

On the other hand, if we window first and then shift (or, as I suggested, delay in the frequency domain), then we still get tapering at the frame boundaries, but the phase is computed relative to the window center rather than the frame start.

**Comment by OverLordGoldDragon** [comment]
That's just simple code for those willing to try on their local `librosa`s; the goal with `frame(modulated=True)` will be to achieve `ifftshift(frame(modulated=False))` upon very first write, so changing indexing/stride logic etc.

**Comment by bmcfee** [comment]
> That's just simple code for those willing to try on their local `librosa`s; the goal with `frame(modulated=True)` will be to achieve `ifftshift(frame(modulated=False))` upon very first write, so changing indexing/stride logic etc.

Ok, thanks for clarifying.  However, I think the rest of my previous comment still stands: shifting only the window does not achieve the desired outcome.

**Comment by OverLordGoldDragon** [comment]
> shifting only the window does not achieve the desired outcome.

Right, didn't mean to imply otherwise.

**Comment by OverLordGoldDragon** [comment]
Implemented it, including with GPU acceleration, [here](https://github.com/OverLordGoldDragon/ssqueezepy/blob/2fc8c91a93943d432b0ba889e2cf080851641786/ssqueezepy/utils/stft_utils.py#L18), tested [here](https://github.com/OverLordGoldDragon/ssqueezepy/blob/2fc8c91a93943d432b0ba889e2cf080851641786/tests/fft_test.py#L380).

**Comment by bmcfee** [comment]
Circling back on this; #1351 includes a fix for #1375 , in which we found an implementation error in `util.frame` when the input has `ndim>2`.  The rewrite in #1351 avoids copying in all cases, and creates a read-only view of the data.  This ultimately precludes anything like a "modulated" parameter to `frame`.

If we implement this feature, it will have to be done as a post-processing step on the stft, as I've advocated previously.
